extensions/wikia/CategoryPagination/CategoryPaginationViewer.class.php

```diff
@@ -2,102 +2,230 @@
+
+	/**
+	 * @var Paginator[] paginator object for each of the sections
+	 */
 	private $paginators = [];
 
-	public function __construct( $title, IContextSource $context, $page ) {
-		parent::__construct( $title, $context );
+	/**
+	 * @var CategoryTree
+	 */
+	private $tree = null;
+
+	/**
+	 * Get the category page HTML
+	 *
+	 * @return string
+	 */
+	public function getHTML() {
+		global $wgCategoryMagicGallery, $wgEnableCategoryTreeExt, $wgCategoryTreeCategoryPageOptions;
 
-		$this->page = $page;
+		// Image galleries
+		$this->showGallery = $wgCategoryMagicGallery && !$this->getOutput()->mNoGallery;
+		$this->clearCategoryState();
+
+		// Initialize category tree
+		if ( !empty( $wgEnableCategoryTreeExt ) && !$this->getRequest()->getCheck( 'notree' ) ) {
+			$this->tree = new CategoryTree( $wgCategoryTreeCategoryPageOptions );
+			$this->tree->setHeaders( $this->getOutput() );
+		}
+
+		$this->generateData();
+
+		/**
+		 * Hook for Answers to add their section
+		 * @see answerCategoryOtherSection
+		 */
+		$answersSection = '';
+		wfRunHooks( 'CategoryViewer::getOtherSection', [ &$this, &$answersSection ] );
+
+		/**
+		 * Hook for category galleries to add the top pages at the top
+		 * @see CategoryGalleriesHelper::onCategoryPageGetCategoryTop
+		 */
+		$categoryGallery = '';
+		wfRunHooks( 'CategoryPage::getCategoryTop', [ $this, &$categoryGallery ] );
+
+		// Render
+		$this->getOutput()->addHeadItem( 'Paginator', $this->getPaginationHeadItem() );
+		$tpl = new EasyTemplate( __DIR__ . '/templates' );
+		$tpl->set_vars( [
+			'aswersSection' => $answersSection,
+			'categoryGallery' => $categoryGallery,
+			'content' => $this->content,
+			'counts' => $this->counts,
+			'empty' => empty( $answersSection ) && empty( $this->content ),
+			'lang' => $this->getLanguage(),
+			'numberShown' => $this->numberShown,
+			'paginator' => $this->paginators,
+			'titleEscaped' => htmlspecialchars( $this->getTitle()->getText() ),
+		] );
+		return $tpl->render( 'category' );
+	}
 
-		$cat = Category::newFromTitle( $this->title );
+	/**
+	 * Generate the data for each of the default sections
+	 * Save it to $this->content and $this->numberShown
+	 */
+	private function generateData() {
+		// Get counts
+		$cat = $this->getCat();
 		$this->counts = [
-			'file' => $cat->getFileCount(),
-			'subcat' => $cat->getSubcatCount(),
-			'page' => $cat->getPageCount() - $cat->getFileCount() - $cat->getSubcatCount(),
+			self::TYPE_FILE => $cat->getFileCount(),
+			self::TYPE_SUBCAT => $cat->getSubcatCount(),
+			self::TYPE_PAGE => $cat->getPageCount() - $cat->getFileCount() - $cat->getSubcatCount(),
 		];
 
-		$url = $this->title->getLocalURL();
-
-		foreach ( array_keys( $this->from ) as $type ) {
-			$this->from[$type] = $this->getActualFrom( $type, $page );
+		// Set up pagination
+		$url = $cat->getTitle()->getFullURL();
+		$this->page = max( 1, $this->getRequest()->getInt( 'page', 1 ) );
+		foreach ( array_keys( $this->counts ) as $type ) {
```

Comment:
foreach ($this->counts as $type => $value) {
	$this->paginators[$type] = new Paginator( $value, $this->limit, $url );
	$this->paginators[$type]->setActivePage( $this->page );
}
?